graph/patterns/change-status.md

```diff
@@ -0,0 +1,98 @@
+
+## Problem
+
+In some cases, exposed resources can trigger side effects that modify the system behavior. For instance, a [`riskyUser`](https://docs.microsoft.com/en-us/graph/api/resources/riskyuser?view=graph-rest-1.0) can be [`confirmed`](https://docs.microsoft.com/en-us/graph/api/riskyuser-confirmcompromised?view=graph-rest-1.0&tabs=http) or [`dismissed`](https://docs.microsoft.com/en-us/graph/api/riskyuser-confirmcompromised?view=graph-rest-1.0&tabs=http).
+
+These situation are often modeled as a `POST` request to a specific endpoint in the same URI space of the resource, triggering the side effect directly and not returning any data. While it might be a quick and easy solution, there are some considerations:
+
+- The API should be exposing resources and allow their manipulation, while Actions resemble a remote procedure call on an object, typical of the[OOP][1]
+- It couples the request with the side effect, making more difficult to test, log, replay and reason about what is happening in the system
+- It forces the consumer to follow the [OOP][1] approach at the API level. We do not want to pursue any specific style and the developers writing the client might be working in other ways
+
+## Solution
+
+API Designers can prefer the usage of a `status` property or a datafyed version of the intended action to represent the **intent** of triggering a side effect rather than starting the execution directly through the HTTP request.
+
+Taking in consideration the example of the `riskyUser` above, it is possible to add a `status` property to the resource:
+
+```xml
+<EnumType Name="riskyStatus">
+    <Member Name="confirmed" Value="0"/>
+    <Member Name="dismissed" Value="1"/>
+    <Member Name="potential" Value="2"/>
+</EnumType>
+
+<EntityType Name="riskyUser" >
+    <Property Name="status" Type="riskyStatus" />
+</EntityType>
+```
+
+To confirm the user as compromised, we can now send a PATCH request on the resource:
+
+```bash
+curl -X PATCH --json '{"status": "confirmed"}' https://graph.microsoft.com/v1.0/riskyUsers/a57dc75f-24b5-47ce-b5e1-44822f5d4729
+```
+
+Instead of returning an empty body with `204` status code, depending on the situation, the response should contain the status of the request, if can be satisfied in a reasonable time:
+
+```json
+{
+  "status": "confirmed"
+}
+```
+
+If the response triggers a long running operation, the response should contain a reference to query its status, following the [existing guidance][2]:
+
+```json
+{
+  "createdDateTime": "2015-06-19T12-01-03.4Z",
+  "status": "disabling"
+}
+```
+
+The status on the resource shall stay the same until the long running operation is completed, but the workload can optionally return a `desiredStatus` property along with the other resource's properties to signal that it is working on a request:
+
+curl https://graph.microsoft.com/v1.0/riskyUsers/a57dc75f-24b5-47ce-b5e1-44822f5d4729
+
+```json
+{ "status": "potential", "desiredStatus": "confirmed" }
```

Comment:
I think status here should be the existing riskyState.  Not sure about the "potential" value though - it should probably show the current state until it's updated.  The object also has a riskLastModifiedDateTime, and riskDetail that would both be updated as a side-effect of the riskyState updating.

Reply:
> I think status here should be the existing riskyState

Isn't it the case? We are asking to return back the current status for the property - in this case the `riskyState = potential`

>  it should probably show the current state until it's updated

Isn't it what the guideline is saying?

> The status on the resource shall stay the same until the long running operation is completed

---

> The object also has a riskLastModifiedDateTime, and riskDetail that would both be updated as a side-effect of the riskyState updating

Good point. Maybe we can change the guideline to say "you should also return all the appropriate properties changed because of the request of change of status", what do you think?